From cc3ed12cae3fd8969c79b5270f43aea7aef282fa Mon Sep 17 00:00:00 2001
From: Nicolin Chen <b42378@freescale.com>
Date: Tue, 13 Aug 2013 11:37:54 +0800
Subject: [PATCH 0377/5242] MLK-11479-04 ASoC: imx-wm8962: Set MCLK source
 clock to 0Hz in hw_free()

commit  6e49dddf3b726608d4a8bbf012dfdd9533c3db37 from
https://source.codeaurora.org/external/imx/linux-imx.git

cherry-pick below patch from v3.14.y:
ENGR00274386-2 ASoC: imx-wm8962: Set MCLK source clock to 0Hz in hw_free()

When DAPM closed WM8962 after playback, its driver would prompt
'wm8962 0-001a: Unsupported sysclk ratio 500' due to the invalid
divisor calculated by WM8962 codec driver.

To fix it, we can work around by setting its MCLK source to 0Hz,
so the codec driver would never get an invalid divisor any more.
Since hw_params() would re-set the MCLK source, no need to worry
about any side-effect.

Acked-by: Wang Shengjiu <b02247@freescale.com>
Signed-off-by: Nicolin Chen <b42378@freescale.com>
(cherry picked from commit a935f7680ac3958ce72cf7413cac278c0683d4c0)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/imx-wm8962.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sound/soc/fsl/imx-wm8962.c b/sound/soc/fsl/imx-wm8962.c
index 30dd376..aad4974 100644
--- a/sound/soc/fsl/imx-wm8962.c
+++ b/sound/soc/fsl/imx-wm8962.c
@@ -129,7 +129,7 @@ static int imx_hifi_hw_free(struct snd_pcm_substream *substream)
 		 * So we set MCLK as sysclk once, which'd remove the limitation.
 		 */
 		ret = snd_soc_dai_set_sysclk(codec_dai, WM8962_SYSCLK_MCLK,
-				data->clk_frequency, SND_SOC_CLOCK_IN);
+				0, SND_SOC_CLOCK_IN);
 		if (ret < 0) {
 			dev_err(dev, "failed to switch away from FLL: %d\n", ret);
 			return ret;
-- 
1.7.9.5

